walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,92 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.List;
+import java.util.Stack;
 
 public final class FindMeetingQuery {
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<TimeRange> ret = new ArrayList<>();
+    if (request.getDuration() <= 0) {
+      return ret;
+    }
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return ret;
+    }
+    List<TimeRange> occupiedBlocks = recordOccupiedBlocks(events, request.getAttendees());
+    ret = complement(TimeRange.WHOLE_DAY, occupiedBlocks);
+    ret.removeIf(tr -> tr.duration() < request.getDuration());
+    return ret;
+  }
+
+  /* Return list of blocks occupied by attendees, unordered */
+  private List<TimeRange> recordOccupiedBlocks(Collection<Event> events, Collection<String> names) {
+    List<TimeRange> occupiedBlocks = new ArrayList<>();
+    for (Event event : events) {
+      if (isOccupied(event, names)) {
+        occupiedBlocks.add(event.getWhen());
+      }
+    }
+    return occupiedBlocks;
+  }
+
+  /* Return true if anyone in the targetNames is involved in the event */
+  private boolean isOccupied(Event event, Collection<String> targetNames) {
+    for (String targetName : targetNames) {
+      if (event.getAttendees().contains(targetName)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /* Return a list of the available TimeRanges */
+  private List<TimeRange> complement(TimeRange wholeDay, List<TimeRange> mergedTimeRanges) {
```

Comment:
After reading this method without reading the helper method, I didn't understand how it was computing the complement. Could you add a comment above this method explaining the theory on how it is getting computed?

Reply:
done